readme.txt
Vortex Tracker II v1.0 beta 20
(c)2000-2022 S.V.Bulba
Author Sergey Bulba
E-mail: svbulba@gmail.com
Support page: http://bulba.untergrund.net/

History
-------

The Vortex Tracker II idea is based on the Vortex Tracker. I and Roman
Scherbakov have started Vortex Tracker project at summer 2000, and the work was
stopped at autumn 2000 in not finished state. At August 2002 I decided to
restart work alone with new name Vortex Tracker II.

What is it?
-----------

Vortex Tracker II is a complete music editor for creating and editing music for
AY-3-8910, AY-3-8912 or YM2149F sound chips. Sound output is realized by sound
chips emulation through standard Windows digital sound devices, so, real sound
chips are not required. Vortex Tracker II uses standard Win32 functions and does
not require any additional libraries.

Vortex Tracker II can import ZX Spectrum music files (modules) of next types:

 1) Pro Tracker 3.xx (file mask is *.pt3);
 2) Pro Tracker 2.xx (*.pt2);
 3) Pro Tracker 1.xx (*.pt1);
 4) Flash Tracker (*.fls);
 5) Fast Tracker (*.ftc);
 6) Global Tracker 1.x (*.gtr);
 7) Pro Sound Creator 1.xx (*.psc);
 8) compiled  Pro Sound Maker modules (*.psm);
 9) compiled ASC Sound Master 0.xx modules (*.as0);
 10) compiled ASC Sound Master 1.xx modules (*.asc);
 11) uncompiled Sound Tracker modules (*.st1);
 12) compiled Sound Tracker and Super Sonic modules (*.stc);
 13) compiled Sound Tracker 3 (*.st3);
 14) uncompiled Sound Tracker Pro modules (*.stf);
 15) compiled Sound Tracker Pro modules (*.stp);
 16) compiled SQ-Tracker modules (*.sqt);
 17) Amadeus (Fuxoft AY Language) modules (*.fxm, *.ay).

VT II detects module type only by filename extension (mask), and no any
additional checks are performed. These extensions are used in well-known player
called Ay_Emul. Any other extensions are analyzed as text files.

PT v3.7+ modules saved in TS-mode can be imported in VT II too, they are
converted to two single PT3 in two windows. Previous not documented PT v3.6 TS-
modules are imported after user prompt.

Any two tracker modules (except FXM) can be stored in one file with 16 bytes
identifier at the end of file. For text format identifier is not need. After
loading such pair VT II turns on Turbo-Sound mode and aligns both windows
horizontally.

Vortex Tracker II saves result only in one format is Pro Tracker 3.xx (*.pt3).
You can play these files in different players-emulators (the most known is
Windows and Linux emulator Ay_Emul), on real ZX Spectrum in many players
(Little Viewer, Quick Commander, Real Commander, BestView, Pusher, ZXAmp, The
Viewer and so on) or by built-in playing routines. Also you can include
YM-sound into your PC-programs (see Ay_Emul sources, YM-Engine or SquareTone at
http://bulba.untergrund.net/).

During editing you can save work versions of module in text format. It allows
to save all temporary not used ornaments, samples and patterns. Also, text
format is easy editable in any text editor. Of course, text format is only
one chance to save your music, if PT3 saving is not available due size
limitations (65536 bytes).

In Turbo-Sound mode during saving one of module second module is saved to the
end of file too.

In fact, Vortex Tracker II is PC version of ZX Spectrum Pro Tracker 3.xx.
The most compatible version is Pro Tracker v3.6x-3.7x of Alone Coder (a.k.a.
Dima Bystrov). Vortex Tracker II is fully compatible with any Pro Tracker v3.5x
in "ProTracker 3.5" mode.

All supported formats are converted to Pro Tracker 3 compatible format,
therefore some information can be lost, because of ZX music formats are very
badly compatible between each other. More information about converting you can
see in 'Tracker limitations.rus.txt' file.

New 3xxx interpretation is supported in last Pro Tracker 3 versions from Alone
Coder (v3.6 and higher).

During editing total time length and current time position are calculated
automatically (for demomakers).

Note: new 3xxx interpretation changes behavior of ASC modules import also.

This version has next new features:

06/17/2022:

1.Text module loader slightly extended to load Ivan Pirog's vt2, vtp, vts and vto.

06/16/2022:

1. Windows 9x is not supported now.
2. Bug fix: right length in hobeta header for non-TS module without player (was
   16 bytes bigger).
3. Digital sound engine changed to Ay_Emul 2.9 beta 31's.
4. More perfect AY emulation engine from Ay_Emul 2.9 beta 31.
5. Options are stored in VT.cfg now (like Ay_Emul in home folder of current
   user, or in VT executable folder, if you move it or create empty yourself).
6. STC loader can extract author' strings from 'KSA SOFTWARE COMPILATION OF '
   and 'SOUND TRACKER COMPILATION OF ', insterted in Ay_Emul methodic.
7. Added import of uncompiled Sound Tracker 1.xx (ST1 file extension).
8. Added import of compiled Sound Tracker 3 (ST3 file extension).
9. Added import of uncompiled Sound Tracker Pro (STF file extension).
10.Added import of compiled ASC Sound Master 0 (AS0 file extension).
11.TS-module's windows are aligned after opening even if other modules was
   loaded bedore.
12.In windows loop play mode added checking if TS-pair to skip replaying twice.
13.If TS-window playing starts, linked TS-window brings to foreground now.

Known problems
--------------

See readme.rus.txt

Short manual
------------

Next keys are used.

 Pattern editor keys
 ~~~~~~~~~~~~~~~~~~~

1. In a note cell

1.1. NoteKeys:

                          Q 2 W 3 E R 5 T 6 Y 7 U I 9 O 0 P [ = ]
  Z S X D C V G B H N J M , L . ; /

  | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | |
  | C#| D#| | F#| G#| A#| | C#| D#| | F#| G#| A#| | C#| D#| | F#|
  C   D   E F   G   A   B C   D   E F   G   A   B C   D   E F   G

So fortepiano keys are emulated with range more than two octaves.

- If cursor in a note cell of track column, use numpad keys from 1 to 8 to
 (only when NumLock is on), key "A" for (R--) and key "K"  (---).

1.2. Shift+NoteKeys - input note one octave higher.

1.3. Shift+Ctrl+NoteKeys - input note one octave lower.

1.4. A - for sound off in channel command (R--).

1.5. K - for delete note (---).

1.6. From 1 to 8 at NumPad - choose current octave number.

During inputting numbers from NumPad NumLock must be turned on.

2. Pattern navigation (cursor control)

2.1. Up, Down, Left, Right - move cursor in four directions.

2.2. PageUp, PageDown - move cursor to page up or down.

2.3. Home, End - move cursor to begin or end of line.

2.4. Ctrl+Right, Ctrl+Left - quick cursor moving among columns.

2.5. Ctrl+PageUp, Ctrl+PageDown - move cursor to top or bottom of column.

2.6. Ctrl+Home, Ctrl+End - move to top of first or bottom of last column.

2.7. Mouse clicks - move cursor to desired cell.

2.8. Mouse wheel - scroll pattern up/down.

3. Pattern's area selection

3.1. Shift+cursor control keys (i.2.) - select rectangular area of pattern.

3.2. Ctrl+A, Ctrl+5 on NumPad - select all pattern.

3.3. Moving mouse with pushed left button.

3.4. Shift+mouse click - select from cursor to desired cell.

3.5. Shift+mouse wheel - not reset/add to selection.

4. Jumps from edit pattern to other objects

4.1. Tab - jump from one window control to other in predefined order.

4.2. Shift+Tab - same in reverse order.

4.3. ` (back apostrophe, usually over Tab key) - quick switch between pattern
and position list editors.

5. Deleting

5.1. BackSpace - delete current track line (with moving).
                                                          
5.2. Ctrl+BackSpace, Ctrl+Y - delete current pattern line (with moving).

5.3. Delete - clear selection's values or value at cursor position.

5.4. Ctrl+Delete - clear pattern line values.

5.5. Numpad / - delete each 2nd pattern line with moving and decreasing pattern
     size.

6. Inserting empty lines

6.1. Insert - insert empty track line (with moving).

6.2. Ctrl+I - insert empty pattern line (with moving).

6.3. Numpad * - double pattern size with inserting empty pattern lines between
     old lines.

7. Working with selection

7.1. Ctrl+C, Ctrl+Insert - copy selection to clipboard.

7.2. Ctrl+X, Shift+Delete - same with clearing selection values.

7.3. Ctrl+V, Shift+Insert - inserting from clipboard (to right and to down from
     cursor or from upper left corner of selection). In case of selection is
     defined, insertion not exceeds its bounds.

7.4. Ctrl+Alt+V - same as usual inserting from clipboard, but without replacing
     non-empty values with empty ones (i.e. merging).

7.5. Numpad + and Numpad - - transpose 1 semitone up and down.

7.6. Ctrl+Numpad + and Ctrl+Numpad - - transpose 1 octave up and down.

8. Test playing of pattern's part

8.1. Any pushed key during inputting data - play current line.

8.2. Pushed Enter key - play pattern from current cursor position until key
is pushed. If any looping is on, pattern will be playing cyclically,

9. Other

9.1. 0 on NumPad - on/off Auto Envelope.

9.2. Space - on/off Auto Step.

9.3. Shift+Backspace - if Auto Step is on do step backward.

Right mouse button can call pop-up menu, which duplicates some key combinations
(useful if you are using keyboard without Numpad, like in notebooks).

 Common keys
 ~~~~~~~~~~~

1. Play controls

1.1. F5 - play module from current position.

1.2. F6 - play module from start.

1.3. F7 - play current pattern from current line.

1.4. F8 - play current pattern from start.

1.5. Esc - stop playing and go to edit pattern; also use to free sound device.

1.6. Ctrl+L - module and pattern loop playing on/off.

1.7. Ctrl+Alt+L - loop playing among all opened modules on/off. If only one
module is opened or pattern is playing, it works as usual loop.

2. Sound chip emulation

2.1. Ctrl+Alt+C - toggle chip type (AY-3-8910/12 or YM2149F).

2.2. Ctrl+Alt+A - toggle channel allocation (Mono, ABC, ACB and BAC, all other
can be selected in 'Options' dialog).

3. Editing

3.1. Ctrl+E - toggle autoenvelope mode to autocalculating envelope period by its
type and current note.

3.2. Ctrl+R - toggle autostep (edit spacing).

3.3. Ctrl+T - call Track Manager dialog.

3.4. Ctrl+Alt+T - call Global Transposition dialog.

4. Standard keys

4.1. Alt+F4 - close Vortex Tracker II.

4.2. Ctrl+F4 - close active window with module.

4.3. Ctrl+F6 - cyclic choosing of opened modules.

4.4. Ctrl+O - call open dialog.

4.5. Ctrl+S - save module.

4.6. Tab, Shift+Tab - cyclical jumps between window controls (forward and
backward).

4.7. Alt+BackSpace - undo last change.

4.8. Alt+Enter - redo last change.

 Positions list editor keys
 ~~~~~~~~~~~~~~~~~~~~~~~~~~

1. ` - jump to/from pattern editor.

2. Left, Right or left mouse button - select position (during playing selected
position will restart playing).

3. Right mouse button - select position and call popup menu.

4. L - set loop position.

5. From 0 to 9 - enter pattern number for selected position.

6. Del - delete position with moving.

7. Ins - insert position with moving.

- In the sample editor:

Up, Down, Left, Right, PageUp, PageDown, Home, End, Ctrl+Right, Ctrl+Left,
Ctrl+PageUp, Ctrl+PageDown, Ctrl+Home, Ctrl+End for navigation.

In any position of editing sample:

T - toggle on/off tone mask
N - toggle on/off noise mask
M - toggle on/off envelope mask
Alt+Right - add current sample line into sample line templates list
Alt+Left - copy selected line from sample line templates into current sample
	   line.

In 'TNE' columns:
Space - toggle on/off corresponding mask

In any '+' and '-' columns:
Space - toggle sign
Shift+'=', '=', Numpad '+' - change sign to '+'
'-', Numpad '-' - change sign to '-'
Shift+6 ('^') - turning on accumulation in corresponding column '^'
Shift+'-' ('_') - turning off accumulation in corresponding column '_'
0-9,A-F - enter hexadecimal numbers

In any '^' and '_' columns:
Space - on/off accumulation
Shift+6 ('^') - turning on accumulation '^'
Shift+'-' ('_') - turning off accumulation '_'
0-9,A-F - enter hexadecimal numbers

In last column (volume control) '+', '-' and '_':
Space - toggle three variants
Shift+'-' ('_') - don't change sample volume '_'
'-',  Numpad '-' - decrease sample volume by one '-'
'+',  Numpad '+' - increase sample volume by one '+'
0-9,A-F - enter hexadecimal numbers

In the number fields:
0-9,A-F - enter hexadecimal numbers
Space - change sign, in amplitude column - volume control
Shift+'=', '=', Numpad '+' - change sign to '+'
'-', Numpad '-' - change sign to '-'
Shift+6 ('^') - turning on accumulation '^'
Shift+'-' ('_') - turning off accumulation '_'

Any non-digital key or mouse clicking reset number inputting counter.

Right mouse clicking in amplitude visualization field ('*' symbols) to choose
corresponding amplitude.

Moving mouse with clicked right button in amplitude visualization field to
draw amplitude.

Right mouse click in some sample cells to toggle corresponding value.

- In the ornament editor:

Up, Down, Left, Right, PageUp, PageDown, Home, End, Ctrl+PageUp, Ctrl+PageDown,
Ctrl+Home, Ctrl+End for navigation

0-9 - input decimal numbers

Space - toggle number sign

Shift+'=', Numpad '+' - set '+' sign

'-', Numpad '-' - set '-' sign

Use right mouse click to toggle sign of corresponding value.

- In sample and ornament test fields navigation and editing same as in patterns
editor.

Maximum length of title and author strings is 32 chars.

Highlight step for pattern lines can be adjusted in pattern window. If Auto is
on, step is autoselected from 3,4 and 5 depending of pattern length.

You can load VT II modules into ZX Spectrum Pro Tracker 3.69x-3.7x. It can load
modules with VT II header and with *.pt3 file extension.

For playing module by standard ZX Spectrum playing routine don't select song
speed smaller than 2 (3 for old players) or better use VT II built-in player
without such limitations.

To moral support author of VT II, select new Vortex Tracker II header, please.

Save command available only after starting editing of song.

Header types
------------

Current version can save three header types in PT3-file.

1) 'Vortex Tracker II 1.0 module: '

This header means, that module can contain all VT II abilities, including new
3xxx interpretation. However, module can be played with old PT3-players without
any problems in most cases.


History.txt
Vortex Tracker II v1.0 beta 10

12/19/2004:

1. New Vortex Tracker II v1.0 PT3 player for ZX Spectrum Release 6 is supported.

01/02/2005:

2. Max length of pattern is 100 lines now. It is compatible with all players
   (on ZX, PC and ST). Don't recompile such modules in original ZX Spectrum
   PT3 editors (there are 64 lines max).

01/03/2005:

3. Undesired area selection during autostep was removed.
4. Highligt step for tracks' lines can be adjusted now. If Auto is on, step is
   autoselected from 3,4 and 5 depending of pattern length (specially for
   Key-Jee).

01/04/2005:

5. Added checking of module size during saving (compiling) PT3 format. If size
   is greater than 65536, error message will appear. If it is, save module in
   text format and than try to solve the problem. Alone Coder proposes to split
   it to several small modules and then play it one by other. So, you can load
   module in several windows, remove some data and save them with different
   names.
6. Added checking of block size during export to ZX-formats.

Vortex Tracker II v1.0 beta 11

04/23/2005:

1. Added Amadeus (Fuxoft AY Language) modules import. Both FXM and ZXAYAMAD are
   supported. The most modules require manual editing, but sounds good and
   without it.

04/24/2005:

2. Memory for patterns/samples/ornaments are static now, i.e. you can decrease
   and increase its length without losting last lines previous values.
3. Fixed error: fault when trying to play empty (default) pattern by F7/F8.
4. Patterns/samples/ornaments length can be changed during playing now.

04/26/2005:

5. Maximal length of pattern is 256 lines now. To keep 2 columns line numeration
   it is hexadecimal now.

05/01/2005:

6. Undo/Redo functions are added. Try to test and send me bugreports or wishes.

05/04/2005:

7. Due HalfElf's request, added ability to load modules into editor from command
   line. Several files must be divided by spaces, and filenames with spaces must
   be placed to doublequotes.

Vortex Tracker II v1.0 beta 12

05/29/2005:

1. Fixed error in PT2 importer: memory before sample import was not clear.

06/14/2005:

2. Fixed horrible mistake: if you open save or some other dialogs during playing
   in "Loop among all opened modules" and next window start playing before you
   will finish, dialog will work with new active window. From this version
   all dialogs work with window, which was active during calling dialog.
   
09/16/2005:

3. Added import of PSC v1.00-1.03 modules.

09/30/2005:

4. Added import of compiled PSM (Pro Sound Maker by Denis Dratov) modules. They
   must be compiled by original (author's) compiler.

Vortex Tracker II v1.0 beta 13

10/02/2005:

1. Fixed error in PSM import: loop position always was zero.

03/14/2006:

2. In addition to 'Save As...' command added simple 'Save' (asks for file name
   only for new modules).
3. Added save requests before closing module windows or whole editor.

03/16/2006:

4. Added second sound chip emulation to play any two opened modules
   simultaneously. 'Play' and 'Play from start' start playing module of current
   window through the first chip and module selected in dropdown list of control
   panel through the second chip. So, you can edit two standard PT3-modules, and
   then play it simultaneously through some of known devices (Turbo-Sound or
   Turbo-AY) on real ZX (see example in ZX-magazine InfoGuide #8).

03/18/2006:

5. MC68000 PT3-player is updated (SNDH-export). Note and volume tables are
   packed now via Ivan Roshin methods (player size is 1 KB smaller now).
6. SNDH-export is improved: new SNDH v2 TAGs are supported and new file name
   extension (SNDH) is used now.

03/19/2006:

7. 'Options' dialog allows you to input any AY-frequency in range 1-3.5 MHz, and
   any player frequency in range 1-2000 Hz.
8. Added track colors selection in 'Options' dialog.

03/25/2006:

9. Fixed tracks editor's cursor moving errors (PageUp/PageDown reaction).

Vortex Tracker II v1.0 beta 14

10/01/2006:

1. Window for 2nd chip is autoselected during changing module window. No need to
   select it manually now (if you are editing only two modules in TS mode, of
   course).
2. During playing in TS mode tracks moving of both windows are sinchronized now.
3. During editing in TS mode tracks and cursors of both windows are sinchronized
   too.
4. In TS mode cursor moving to the maximum right or left position of tracks
   editor will activate window of 2nd chip in same position of tracks editor.
5. "Play from position" function (F5) works from current cursor position now.

10/03/2006:

6. Fixed bug: envelope period changed in AutoEnv mode was not added to the undo
   list (thanks to Karbofos for bug-report).

Vortex Tracker II v1.0 beta 15

03/04/2007:

1. Module window is redesigned for more comfortable editing in TS mode (thanks
   to Znahar for ideas and help).
2. No need to close Tracks Manager window now, it stays on top and works with
   current module, which name is duplicated in main window title (specially for
   (Znahar).

03/08/2007:

3. Selected track area can be transposed by calling pop-up menu or from keyboard
   (Num +, Num -, Ctrl+Num +, Ctrl+Num - (+-semiton/+-octave). New function
   works with notes and envelope track.
4. Fixed error: after using AutoStep current time was not updated.
5. AutoStep works after pasting from clipboard now (specially for TAD).
6. Fixed error: trying to start playing of just created (empty) module in TS
   mode rised error message (thanks to Znahar for bug-report).
7. AutoStep can be used with insert/delete/clear track/pattern line. Can be
   useful for increasing/decreasing pattern size twice (thanks to TAD for idea).
8. Fixed error: the most Undo/Redo operations not updated module duration info.

HotKeys.pas
{
This is part of Vortex Tracker II project

Version 2.0 and later
(c)2017-2021 Ivan Pirog, ivan.pirog@gmail.com
https://github.com/ivanpirog/vortextracker
}

unit HotKeys;

interface

uses ActnList, Menus, Classes, Dialogs, ComCtrls, StrUtils, Controls, SysUtils, UiTypes;

const
  HK_PLAY_STOP            = 0;
  HK_PLAY_FROM_LINE       = 1;
  HK_PLAY_FROM_START      = 2;
  HK_PLAY_PATT_FROM_LINE  = 3;
  HK_PLAY_PATT_FROM_START = 4;
  HK_STOP                 = 5;
  HK_LOOP_TOGGLE          = 6;
  HK_LOOP_TOGGLE_ALL      = 7;
  HK_LOOP_SET_POSITION    = 8;
  HK_POSITIONS_INSERT     = 9;
  HK_POSITIONS_DELETE     = 10;
  HK_POSITIONS_DUPLICATE  = 11;
  HK_POSITIONS_CLONE      = 12;
  HK_TOGGLE_SAMPLES       = 13;
  HK_TRACKS_MANAGER       = 14;
  HK_TOGGLE_CHIP          = 15;
  HK_TOGGLE_CHANNELS      = 16;
  HK_UNDO                 = 17;
  HK_REDO                 = 18;
  HK_TRANSPOSE_UP1        = 19;
  HK_TRANSPOSE_DOWN1      = 20;
  HK_TRANSPOSE_UP3        = 21;
  HK_TRANSPOSE_DOWN3      = 22;
  HK_TRANSPOSE_UP5        = 23;
  HK_TRANSPOSE_DOWN5      = 24;
  HK_TRANSPOSE_UP12       = 25;
  HK_TRANSPOSE_DOWN12     = 26;
  HK_EXPAND_PATTERN       = 27;
  HK_COMPRESS_PATTERN     = 28;
  HK_OPTIONS              = 29;
  HK_DUP_NOTE_PARAMS      = 30;
  HK_MOVE_BETWEEN_PATRNS  = 31;
  HK_SPLIT_PATTERN        = 32;
  HK_SWAP_CHANS_LEFT      = 33;
  HK_SWAP_CHANS_RIGHT     = 34;
  HK_JMP_PAT_START        = 35;
  HK_JMP_PAT_END          = 36;
  HK_JMP_LINE_START       = 37;
  HK_JMP_LINE_END         = 38;
  HK_COPY_MODPLUG         = 39;
  HK_COPY_RENOISE         = 40;
  HK_COPY_FAMI            = 41;
  HK_PATTERN_PACKER       = 42;


  SystemHotKeys : array[0..12] of string = (
     'Ctrl+A', 'Ctrl+V', 'Ctrl+N', 'Ctrl+O',
     'Ctrl+C', 'Ctrl+X', 'Ctrl+V', 'Ctrl+W',
     'Ctrl+S', 'Ctrl+Shift+S', 'Shift+Ctrl+S',
     'Alt+Space', 'Alt+F4'
  );

var
  VTHotKeys : array[0..42, 0..1] of string = (
      ('Play/Stop',               'Space'),        // HK_PLAY_STOP
      ('Play from Line',          'F5'),           // HK_PLAY_FROM_LINE
      ('Play Track from Start',   'F6'),           // HK_PLAY_FROM_START
      ('Play Pattern from Line',  'F7'),           // HK_PLAY_PATT_FROM_LINE
      ('Play Pattern from Start', 'F8'),           // HK_PLAY_PATT_FROM_START
      ('Stop',                    'Esc'),          // HK_STOP
      ('Toggle Looping',          'Ctrl+L'),       // HK_LOOP_TOGGLE
      ('Toggle Loopting All',     'Ctrl+Alt+L'),   // HK_LOOP_TOGGLE_ALL
      ('Set Loop Position',       'L'),            // HK_LOOP_SET_POSITION
      ('Insert position',         'Ins'),          // HK_POSITIONS_INSERT
      ('Delete positions',        'Del'),          // HK_POSITIONS_DELETE
      ('Duplicate positions',     'Ctrl+D'),       // HK_POSITIONS_DUPLICATE
      ('Clone positions',         'Ctrl+Shift+D'), // HK_POSITIONS_CLONE
      ('Toggle Samples',          'Ctrl+M'),       // HK_TOGGLE_SAMPLES
      ('Show Tracks Manager',     'Ctrl+T'),       // HK_TRACKS_MANAGER
      ('Toggle Chip',             'Ctrl+Alt+C'),   // HK_TOGGLE_CHIP
      ('Toggle Channels',         'Ctrl+Alt+A'),   // HK_TOGGLE_CHANNELS
      ('Undo',                    'Ctrl+Z'),       // HK_UNDO
      ('Redo',                    'Ctrl+Shift+Z'), // HK_REDO
      ('Transpose +1',            'Num +'),        // HK_TRANSPOSE_UP1
      ('Transpose -1',            'Num -'),        // HK_TRANSPOSE_DOWN1
      ('Transpose +3',            'Shift+Num +'),  // HK_TRANSPOSE_UP3
      ('Transpose -3',            'Shift+Num -'),  // HK_TRANSPOSE_DOWN3
      ('Transpose +5',            'Ctrl+Shift+Num +'), // HK_TRANSPOSE_UP5
      ('Transpose -5',            'Ctrl+Shift+Num -'), // HK_TRANSPOSE_DOWN5
      ('Transpose +12',           'Ctrl+Num +'),       // HK_TRANSPOSE_UP12
      ('Transpose -12',           'Ctrl+Num -'),       // HK_TRANSPOSE_DOWN12
      ('Expand Pattern',          'Ctrl+Shift+/'),     // HK_EXPAND_PATTERN
      ('Compress Pattern',        'Ctrl+Shift+Num *'), // HK_COMPRESS_PATTERN
      ('Options',                 'Ctrl+Shift+O'),     // HK_OPTIONS
      ('Use Last Note params - on/off',  'Shift+Esc'), // HK_DUP_NOTE_PARAMS
      ('Move Between Patterns - on/off', 'Shift+`'),   // HK_MOVE_BETWEEN_PATRNS
      ('Split Pattern',           'Alt+X'),            // HK_SPLIT_PATTERN
      ('Swap selected channels left', 'Ctrl+Alt+Left'),    // HK_SWAP_CHANS_LEFT
      ('Swap selected channels right', 'Ctrl+Alt+Right'),  // HK_SWAP_CHANS_RIGHT
      ('Jump to the pattern first line', 'Ctrl+Home'),     // HK_JMP_PAT_START
      ('Jump to the pattern last line',  'Ctrl+End'),      // HK_JMP_PAT_END
      ('Jump to the line start',         'Home'),          // HK_JMP_LINE_START
      ('Jump to the line end',           'End'),           // HK_JMP_LINE_END
      ('Copy pattern to OpenMPT',        'Ctrl+Alt+C'),    // HK_COPY_MODPLUG
      ('Copy pattern to Renoise',        'Ctrl+Shift+C'),  // HK_COPY_RENOISE
      ('Copy pattern to FamiTracker',    'Alt+C'),         // HK_COPY_FAMI
      ('Pattern packer',                 'Ctrl+P')         // HK_PATTERN_PACKER
  );



procedure InitOptionsHotKeys;
procedure SetDefaultHotKeys;
procedure AssignHotKey(HotKeyIndex: Integer; ShortCutText: string);
procedure LoadHotKeysFromText(HotKeysText: string);
procedure ReAssignHotKey(HotKeyIndex: Integer; ShortCutText: string);
function Split(Delimiter: Char; Str: string): TStrings;
function AllHotKeysToText : string;


implementation

uses options, Main;



procedure InitOptionsHotKeys;
var
  I: Integer;
  ListItem: TListItem;
begin
  with options.Form1.HotKeyList do
  begin
    Clear;
    for I := Low(VTHotKeys) to High(VTHotKeys) do
    begin
      ListItem := Items.Add;
      ListItem.Caption := VTHotKeys[I][0];
      ListItem.SubItems.Add(VTHotKeys[I][1]);
    end;
  end;

end;

procedure SetDefaultHotKeys;
var I : Integer;
begin
  for I := Low(VTHotKeys) to High(VTHotKeys) do
    AssignHotKey(I, VTHotKeys[I][1]);
end;


procedure ReAssignHotKey(HotKeyIndex: Integer; ShortCutText: string);
var I: Integer;
begin

  // If shortcut in SystemHotKeys array (like Ctrl+C, Ctrl+V) - exit
  if AnsiIndexText(ShortCutText, SystemHotKeys) > -1 then
  begin
    ShowMessage('Error: Key "' + ShortCutText + '" is a system key.');
    Exit;
  end;

  // User can't assign the Shift key for the pattern jumps
  if (HotKeyIndex in [HK_JMP_PAT_START, HK_JMP_PAT_END, HK_JMP_LINE_START, HK_JMP_LINE_END]) and
     (AnsiPos('Shift', ShortCutText) <> 0) then
     Exit;

  for I := Low(VTHotKeys) to High(VTHotKeys) do
  begin
    if VTHotKeys[I][1] = ShortCutText then
      if MessageDlg('Key "' + ShortCutText + '" already assigned to "' + VTHotKeys[I][0] +'" function. Assign anyway?', mtWarning, mbOKCancel, 0) = mrCancel then
        Exit
      else
      begin
        // Clear old hotkey
        VTHotKeys[I][1] := '';
        AssignHotKey(I, '');
        options.Form1.HotKeyList.Items[I].SubItems[0] := '';

        // Set new hotkey
        VTHotKeys[HotKeyIndex][1] := ShortCutText;
        AssignHotKey(HotKeyIndex, ShortCutText);
        options.Form1.HotKeyList.Items[HotKeyIndex].SubItems[0] := ShortCutText;
        Exit;
      end;
  end;

  VTHotKeys[HotKeyIndex][1] := ShortCutText;
  AssignHotKey(HotKeyIndex, ShortCutText);
  options.Form1.HotKeyList.Items[HotKeyIndex].SubItems[0] := ShortCutText;

end;





procedure AssignHotKey(HotKeyIndex: Integer; ShortCutText: string);
var
  ShortCut : TShortCut;
begin

  if HotKeyIndex > High(VTHotKeys) then Exit;

  VTHotKeys[HotKeyIndex, 1] := ShortCutText;
  ShortCut := TextToShortCut(ShortCutText);

  with MainForm do case HotKeyIndex of
    HK_PLAY_STOP:            begin PlayStop.ShortCut := ShortCut; Play3.ShortCut := ShortCut; end;
    HK_PLAY_FROM_LINE:       begin PlayFromLine.ShortCut := ShortCut; PlayFromLine1.ShortCut := ShortCut; end;
    HK_PLAY_FROM_START:      begin Play1.ShortCut := ShortCut; Play4.ShortCut := ShortCut; end;
    HK_PLAY_PATT_FROM_START: begin PlayPat.ShortCut := ShortCut; Playpatternfromstart1.ShortCut := ShortCut; end;
    HK_PLAY_PATT_FROM_LINE:  begin PlayPatFromLine.ShortCut := ShortCut; Playpatternfromcurrentline1.ShortCut := ShortCut; end;
    HK_LOOP_TOGGLE:          begin ToggleLooping.ShortCut := ShortCut; Togglelooping1.ShortCut := ShortCut; end;
    HK_LOOP_TOGGLE_ALL:      begin ToggleLoopingAll.ShortCut := ShortCut; Toggleloopingall1.ShortCut := ShortCut; end;
    HK_LOOP_SET_POSITION:    Setloopposition1.ShortCut := ShortCut;
    HK_POSITIONS_INSERT:     Insertposition1.ShortCut := ShortCut;
    HK_POSITIONS_DELETE:     Deleteposition1.ShortCut := ShortCut;
    HK_POSITIONS_DUPLICATE:  DuplicatePosition1.ShortCut := ShortCut;
    HK_POSITIONS_CLONE:      ClonePosition1.ShortCut := ShortCut;
    HK_TOGGLE_SAMPLES:       begin Togglesamples1.ShortCut := ShortCut; Togglesamples1.ShortCut := ShortCut; end;
    HK_TRACKS_MANAGER:       begin Tracksmanager1.ShortCut := ShortCut; Tracksmanager1.ShortCut := ShortCut; end;
    HK_TOGGLE_CHIP:          ToggleChip.ShortCut := ShortCut;
    HK_TOGGLE_CHANNELS:      ToggleChanAlloc.ShortCut := ShortCut;
    HK_UNDO:                 Undo.ShortCut := ShortCut;
    HK_REDO:                 Redo.ShortCut := ShortCut;
    HK_TRANSPOSE_UP1:        TransposeUp1.ShortCut := ShortCut;
    HK_TRANSPOSE_DOWN1:      TransposeDown1.ShortCut := ShortCut;
    HK_TRANSPOSE_UP3:        TransposeUp3.ShortCut := ShortCut;
    HK_TRANSPOSE_DOWN3:      TransposeDown3.ShortCut := ShortCut;
    HK_TRANSPOSE_UP5:        TransposeUp5.ShortCut := ShortCut;
    HK_TRANSPOSE_DOWN5:      TransposeDown5.ShortCut := ShortCut;
    HK_TRANSPOSE_UP12:       TransposeUp12.ShortCut := ShortCut;
    HK_TRANSPOSE_DOWN12:     TransposeDown12.ShortCut := ShortCut;
    HK_EXPAND_PATTERN:       ExpandTwice1.ShortCut := ShortCut;
    HK_COMPRESS_PATTERN:     Compresspattern1.ShortCut := ShortCut;
    HK_OPTIONS:              Options1.ShortCut := ShortCut;
    HK_DUP_NOTE_PARAMS:      DuplicateLastNoteParams.ShortCut := ShortCut;
    HK_MOVE_BETWEEN_PATRNS:  MoveBetwnPatrns.ShortCut := ShortCut;
    HK_SPLIT_PATTERN:        Splitpattern1.ShortCut := ShortCut;
    HK_STOP:                 begin Stop.ShortCut := ShortCut; Stop2.ShortCut := ShortCut; end;
    HK_SWAP_CHANS_LEFT:      SwapChannelsLeft1.ShortCut := ShortCut;
    HK_SWAP_CHANS_RIGHT:     SwapChannelsRight1.ShortCut := ShortCut;
    HK_JMP_PAT_START:        JmpPatStartAct.ShortCut := ShortCut;
    HK_JMP_PAT_END:          JmpPatEndAct.ShortCut := ShortCut;
    HK_JMP_LINE_START:       JmpLineStartAct.ShortCut := ShortCut;
    HK_JMP_LINE_END:         JmpLineEndAct.ShortCut := ShortCut;
    HK_COPY_MODPLUG:         begin CopyToModplugAct.ShortCut := ShortCut; CopyToModplug.ShortCut := ShortCut; end;
    HK_COPY_RENOISE:         begin CopyToRenoiseAct.ShortCut := ShortCut; CopyToRenoise.ShortCut := ShortCut; end;
    HK_COPY_FAMI:            begin CopyToFamiAct.ShortCut := ShortCut; CopyToFami.ShortCut := ShortCut; end;
    HK_PATTERN_PACKER:       PackPatternAct.ShortCut := ShortCut;
  else
  end;
end;


function AllHotKeysToText : string;
var
  I: Integer;
begin
  for I := Low(VTHotKeys) to High(VTHotKeys) do
  begin
    Result := Result + VTHotKeys[I][1];
    if I < High(VTHotKeys) then Result := Result + ',';
  end;
end;


procedure LoadHotKeysFromText(HotKeysText: string);
var
  HotKeysList : TStrings;
  I : Integer;

begin
  HotKeysList := Split(',', HotKeysText);
  for I := 0 to HotKeysList.Count-1 do
  begin
    AssignHotKey(I, HotKeysList[I]);
  end;
  FreeAndNil(HotKeysList);
end;


function Split(Delimiter: Char; Str: string): TStrings;
var
  i: Integer;
  row: string;
begin
  Result := TStringList.Create;
  i := 1;
  row := '';
  while i <= Length(Str) do
  begin
    if (Str[i] = Delimiter) then
    begin
      Result.Add(row);
      row := '';
    end
    else
      row := row + Str[i];
    i := i + 1;
  end;
  if (row <> '') then
    Result.Add(row);
end;


end.
